src/shared/parse-helpers.c

```diff
@@ -248,3 +248,35 @@ int config_parse_path_or_ignore(
 finalize:
         return free_and_replace(*s, n);
 }
+
+int parse_machine_spec(const char *spec, const char **ret_machine, char **ret_user) {
+        _cleanup_free_ char *user = NULL;
+        const char *machine;
+
+        assert(spec);
+        assert(ret_machine);
+
+        if (!machine_spec_valid(spec))
+                return -EINVAL;
```

Comment:
frankly, i am not a fan of functions that basically parse twice, once for validation and once for returning the data. But htat's kinda what you do here.

I'd prefer if instead machine_spec_valid() could become a tiny static inline wrapper around parse_machine_spec() that just passes NULL to both return params. And then it is parse_machine_spec() which validates properly